src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs

```diff
@@ -131,11 +131,49 @@ protected override void CheckInterfaces(DiagnosticBag diagnostics)
                 var conversions = new TypeConversions(corLibrary);
                 var location = singleDeclaration.NameLocation;
 
-                foreach (var @interface in interfaces)
+                foreach (var baseInterface in interfaces)
                 {
-                    @interface.CheckAllConstraints(conversions, location, diagnostics);
+                    baseInterface.CheckAllConstraints(conversions, location, diagnostics);
+                }
+
+                // we check direct base interfaces that have the InternalImplementationOnly attribute,
+                // giving an error when such an attributed interface appears in the interface list.
+                foreach (var baseInterface in this.InterfacesNoUseSiteDiagnostics(null))
+                {
+                    EnsureExtensible(baseInterface, location, diagnostics);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Produce an error if the given base interface may not be extended by this type due to the
+        /// presence of the attribute System.Runtime.CompilerServices.InternalImplementationOnlyAttribute
+        /// </summary>
+        private void EnsureExtensible(NamedTypeSymbol baseInterface, SourceLocation location, DiagnosticBag diagnostics)
+        {
+            if (this.ContainingAssembly != baseInterface.ContainingAssembly &&
+                HasInternalImplementationOnlyAttribute(baseInterface) &&
+                !this.ContainingAssembly.HasInternalAccessTo(baseInterface.ContainingAssembly))
+            {
+                diagnostics.Add(ErrorCode.ERR_InternalImplementationOnly, location, this, baseInterface);
+            }
+        }
+
+        /// <summary>
+        /// Return true of the given interface type contains an attribute of the
+        /// name System.Runtime.CompilerServices.InternalImplementationOnlyAttribute
+        /// </summary>
+        private static bool HasInternalImplementationOnlyAttribute(NamedTypeSymbol baseInterface)
+        {
+            foreach (var att in baseInterface.GetAttributes())
+            {
+                if (att.IsTargetAttribute("System.Runtime.CompilerServices", "InternalImplementationOnlyAttribute"))
```

Comment:
The approach you suggest would have the following performance impact:

1. Every attribute in source on a type would now have an additional check for this special attribute (in the sequence of in-then-else's in SourceNamedTypeSymbol.DecodeWellKnownAttribute) as well as an additional field (indirectly through the decoded well known attribute data) whenever the type is checked for any well known attribute. I believe that the situations where it is needed (inheriting from an interface in another assembly that has an annotation) is more rare than the situations where we would pay these costs (every annotation on every type). So there is a time and space penalty compared to the implemented approach (for source types). I'd prefer to use the current approach for source types.
2. For PE symbols, I like your approach but without caching since, again, the situation where the attribute is needed is rare.